perf(ast/estree): pre-allocate `CodeBuffer` for JSON output

overlookmotel opened the pull request:
`ESTree` serialization pre-allocate capacity in `CodeBuffer` for JSON, based on an heuristic ratio of source text length to JSON length, to avoid the buffer having to grow during serialization.

overlookmotel commented:
*Now* we're talking!

Boshen commented:
### Merge activity

* **Feb 24, 11:35 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(ast/estree): pre-allocate `CodeBuffer` for JSON output (#9340)

`ESTree` serialization pre-allocate capacity in `CodeBuffer` for JSON, based on an heuristic ratio of source text length to JSON length, to avoid the buffer having to grow during serialization.
